RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,315 @@
+            int64_t n_rows,
+            int64_t n_cols,
+            TrigDistName tn = TrigDistName::Hadamard
+        ) : n_rows(n_rows), n_cols(n_cols), family(tn) {
+            dim_short = MIN(n_rows, n_cols);
+            dim_long = MAX(n_rows, n_cols);
+}
+};
+
+    template<typename T, SignedInteger sint_t, typename RNG = r123::Philox4x32>
+    struct TrigSkOp {
+        using generator = RNG;
+        using state_type = RNGState<RNG>;
+        using buffer_type = T;
+
+        const int64_t n_rows;
+        const int64_t n_cols;
+        int64_t dim_short;
+        int64_t dim_long;
+        // int64_t n_sampled;
+
+        const TrigDist dist;
+
+
+        const RNGState<RNG> seed_state;
+
+        RNGState<RNG> next_state;
+
+        const blas::Layout layout;
+        const bool sketchFromLeft = true;
+        bool known_filled = false;
+
+        T* DiagonalRademacher = nullptr;
+        T* SampledRows = nullptr;
+
+        TrigSkOp(
+            TrigDist dist,
+            RNGState<RNG> const &state,
+            bool known_filled = true
+        ) : n_rows(dist.n_rows), n_cols(dist.n_cols), dist(dist), seed_state(state), known_filled(known_filled), dim_short(dist.dim_short), dim_long(dist.dim_long){
+            // Memory for Rademacher diagonal gets allocated here
+            // Number of rows/cols to be sampled gets decided here
+            // i.e. `n_sampled` gets set
+            if(sketchFromLeft)
+                DiagonalRademacher = new T[n_rows];
+            else
+                DiagonalRademacher = new T[n_cols];
+
+            SampledRows = new sint_t[dim_short];
+
+            //TODO: Logic to compute the number of samples that we require `r`
+            // this can be shown to depend on the maximal row norm of the data matrix
+            //NOTE: Do not have access to this in the data-oblivious regime --- how
+            // do we get access?
+};
+
+        TrigSkOp(
+            TrigDistName family,
+            int64_t n_rows,
+            int64_t n_cols,
+            uint32_t key
+        ) : n_rows(n_rows), n_cols(n_cols), dist(TrigDist{n_rows, n_cols, family}), seed_state() {};
+
+        ~TrigSkOp();
+};
+
+// Populates the Rademacher diagonal entries and decides the rows/cols to
+// be sampled at the end
+template<typename T, typename RNG, SignedInteger sint_t>
+RandBLAS::RNGState<RNG> fill_trig(
+    TrigSkOp<T, RNG> &Tr
+) {
+    /**
+     * Will do the work of filling in the diagonal Rademacher entries
+     * and selecting the rows/cols to be sampled
+     */
+
+    auto [ctr, key] = Tr.seed_state;
+
+    // Fill in the Rademacher diagonal
+    if(Tr.sketchFromLeft)
+        generateRademacherVector_r123(Tr.DiagonalRademacher, key, ctr, Tr.n_rows);
+    else
+        generateRademacherVector_r123(Tr.DiagonalRademacher, key, ctr, Tr.n_cols);
+
+    //NOTE: Select the rows/cols to be sampled --- use the `repeated_fisher_yates` function
+    int64_t r = Tr.dim_short;
+    int64_t d = Tr.dim_long;
```

Comment:
I'd advise against defining these sorts of temporary variables. Just substitute ``Tr.dim_short`` and ``Tr.dim_long``.